File: code/python/data_structures_article-08.py
"""Monotone stack algorithm template to solve three example problems"""

from __future__ import annotations
from typing import Any


def solve(*args: Any, **kwargs: Any) -> Any:
    """Run this lesson exercise with positional inputs."""
    return _Solution().nextGreaterElement(*args, **kwargs)


class _Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        greater = self.nextGreater(nums2)
        greaterMap = {}
        for i in range(len(nums2)):
            greaterMap[nums2[i]] = greater[i]
        res = []
        for num in nums1:
            res.append(greaterMap[num])
        return res
    def nextGreater(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = [-1] * n
        s = []
        for i in range(n - 1, -1, -1):
            while s and s[-1] <= nums[i]:
                s.pop()
            if s:
                res[i] = s[-1]
            s.append(nums[i])
        return res

File: code/javascript/data_structures_article-08.js
/**
 * Monotone stack algorithm template to solve three example problems
 */
export function solve(...args) {
  // Entry point for the course exercise: pass inputs directly.
  return nextGreaterElement(...args);
}

var nextGreaterElement = function(nums1, nums2) {
    var greater = nextGreater(nums2);
    var greaterMap = new Map();
    for (var i = 0; i < nums2.length; i++) {
        greaterMap.set(nums2[i], greater[i]);
    }
    var res = new Array(nums1.length);
    for (var i = 0; i < nums1.length; i++) {
        res[i] = greaterMap.get(nums1[i]);
    }
    return res;
};
var nextGreater = function(nums) {
    var n = nums.length;
    var res = new Array(n);
    var s = [];
    for (var i = n - 1; i >= 0; i--) {
        while (s.length && s[s.length-1] <= nums[i]) {
            s.pop();
        }
        res[i] = s.length == 0 ? -1 : s[s.length-1];
        s.push(nums[i]);
    }
    return res;
};
